firedrake/parloops.py

```diff
@@ -74,38 +72,50 @@ def _form_kernel(kernel, measure, args, **kwargs):
         else:
             # Do we have a component of a mixed function?
             if isinstance(func, Indexed):
                 c, i = func.ufl_operands
                 idx = i._indices[0]._value
                 ndof = c.function_space()[idx].finat_element.space_dimension()
+                cdim = c.dat[idx].cdim
+                dtype = c.dat[idx].dtype
             else:
-                if len(func.function_space()) > 1:
-                    raise NotImplementedError("Must index mixed function in par_loop.")
-                ndof = func.function_space().finat_element.space_dimension()
+                if func.function_space().ufl_element().family() == "Real":
+                    ndof = func.function_space().dim()  # == 1
+                    kargs.append(loopy.GlobalArg(var, dtype=func.dat.dtype, shape=(ndof,)))
+                    continue
+                else:
+                    if len(func.function_space()) > 1:
+                        raise NotImplementedError("Must index mixed function in par_loop.")
+                    ndof = func.function_space().finat_element.space_dimension()
+                    cdim = func.dat.cdim
+                    dtype = func.dat.dtype
             if measure.integral_type() == 'interior_facet':
                 ndof *= 2
-            if measure is direct:
-                kargs.append(ast.Decl("double", ast.Symbol(var, (ndof,))))
-            else:
-                kargs.append(ast.Decl("double *", ast.Symbol(var, (ndof,))))
-        lkernel = lkernel.replace(var+".dofs", str(ndof))
+            # FIXME: shape for facets [2][ndof]?
+            kargs.append(loopy.GlobalArg(var, dtype=dtype, shape=(ndof, cdim)))
+        kernel_domains = kernel_domains.replace(var+".dofs", str(ndof))
 
-    body = ast.FlatBlock(lkernel)
+    if kernel_domains is "":
+        kernel_domains = "[] -> {[]}"
+    kargs.append(...)
+    knl = loopy.make_function(kernel_domains, instructions, kargs, seq_dependencies=True,
+                              name="par_loop_kernel", silenced_warnings=["summing_if_branches_ops"])
 
-    return pyop2.Kernel(ast.FunDecl("void", "par_loop_kernel", kargs, body),
-                        "par_loop_kernel", **kwargs)
+    return pyop2.Kernel(knl, "par_loop_kernel", **kwargs)
 
 
-def par_loop(kernel, measure, args, kernel_kwargs=None, **kwargs):
+def par_loop(kernel_domains, instructions, measure, args, kernel_kwargs=None, **kwargs):
     r"""A :func:`par_loop` is a user-defined operation which reads and
     writes :class:`.Function`\s by looping over the mesh cells or facets
     and accessing the degrees of freedom on adjacent entities.
 
-    :arg kernel: is a string containing the C code to be executed.
+    :arg kernel_domains: the iteration domain for the kernel, in Integer Set
+        Libarary (ISL) syntax. function_name.dofs can be used as a magic word.
+    :arg instructions: is a string containing the instruction for the kernel
+        in loo.py syntax.
```

Comment:
Can you add some links here?

Reply:
Thinking about it some more, I think that we ought to be able to deduce the common domains automagically, and should provide the option to say what they are.

How about an interface that is something like:

```
# Example nonsense instructions
instructions = """
for i, j, k
    A[i, j] = A[i, j] + A[i, j] * B[i, k]
end
"""
par_loop(instructions, dx, {"A": (a, RW, ("i", "j")), "B": (b, R, ("i", "k")})
```

And we can deduce the index extents from the `dim` of the `dat` in each `Function`, right?

if we optionally want to override the domains, we can say:

```
par_loop(instructions, dx, {...}, domains="{ [i, j, k] : ...}")
```

Achievable?

Reply:
> instructions: is a string containing the instructions

* If feasible, I'd recommend not using strings to pass information to loopy.
* Terminology-wise, I am still thinking of moving to "statement" over "instruction". If it happens, it'll be in a backward-compatible manner, but you may still want to write `statements` rather than `instruction` in your code.

Reply:
Here again, this is meant to be "user-facing". I think it is friendlier for them to write the "string" expression than the equivalent pymbolic, but you could convince me otherwise if it's really easy.